src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java

```diff
@@ -0,0 +1,690 @@
+                    }
+                    return Maybe.empty();
+                },
+            state -> state.backtrackOnFailureMaybe(subState -> {
+                    if (this.unicode) {
+                        return Maybe.empty();
+                    }
+                    Maybe<String> maybeCharacter = subState.nextCodePoint();
+                    if (maybeCharacter.isJust() && !maybeCharacter.fromJust().equals("c")) {
+                        subState.skipCodePoint();
+                        return Maybe.of(maybeCharacter.fromJust().codePointAt(0));
+                    }
+                    return Maybe.empty();
+                })
+        ).apply(superState);
+    }
+
+    private Maybe<Maybe<Integer>> acceptClassEscape(State superState) {
+        return this.<Maybe<Integer>>anyOf(
+            state -> state.backtrackOnFailureMaybe(subState -> {
+                    if (!subState.eat("b")) {
+                        return Maybe.empty();
+                    }
+                    return Maybe.of(0x0008); // backspace
+                }).map(Maybe::of),
+            state -> {
+                    if (this.unicode && state.eat("-")) {
+                        return Maybe.of(Maybe.of((int)"-".charAt(0)));
+                    }
+                    return Maybe.empty();
+                },
+            state -> state.backtrackOnFailureMaybe(subState -> {
+                    if (this.unicode || !subState.eat("c")) {
+                        return Maybe.empty();
+                    }
+                    return subState.eatAny(decimalDigits, new String[]{"_"}).map(str -> str.codePointAt(0) % 32);
+                }).map(Maybe::of),
+            state -> acceptCharacterClassEscape(state) ? Maybe.of(Maybe.empty()) : Maybe.empty(),
+            state -> acceptCharacterEscape(state).map(Maybe::of)
+        ).apply(superState);
+    }
+
+    private Maybe<Maybe<Integer>> acceptClassAtomNoDash(State state) {
+        if (state.eat("\\")) {
+			return this.anyOf(
+					this::acceptClassEscape,
+                subState -> subState.backtrackOnFailureMaybe(subState2 -> {
+						if (!this.unicode && subState2.match("c")) {
+							return Maybe.of(0x005C); // reverse solidus
+						}
+						return Maybe.empty();
+					}).map(Maybe::of)
+			).apply(state);
+        }
+        Maybe<String> nextCodePoint = state.nextCodePoint();
+        if (nextCodePoint.isNothing() || nextCodePoint.fromJust().equals("]") || nextCodePoint.fromJust().equals("-")) {
+            return Maybe.empty();
+        }
+        state.skipCodePoint();
+        return Maybe.of(Maybe.of(nextCodePoint.fromJust().codePointAt(0)));
+    }
+
+    private Maybe<Maybe<Integer>> acceptClassAtom(State state) {
+        if (state.eat("-")) {
+            return Maybe.of(Maybe.of((int)"-".charAt(0)));
+        }
+        return acceptClassAtomNoDash(state);
+    }
+
+    private Maybe<Maybe<Integer>> finishClassRange(State state, Maybe<Integer> atom) {
```

Comment:
Can we have a comment here explaining what each of the three kinds of return value here mean?

Reply:
Ah, never mind, I see you have that below. But ideally move it up to above the first function with that return type.